crypto-sodium/lib/Crypto/Sodium/Salt.hs

```diff
@@ -54,28 +59,64 @@ import qualified Crypto.Sodium.Nonce
 -- of the string you provide. Note that the string will be taken as is,
 -- with all whitespace preserved, for example @[utf8| x |]@ will have length 3.
 utf8 :: QuasiQuoter
-utf8 = QuasiQuoter { quoteExp, quotePat, quoteType, quoteDec }
-  where
-    quoteExp :: String -> Q Exp
-    quoteExp str = [e|unsafeSizedByteArray $(TH.unTypeSplice bsSplice) :: $tSized|]
-      where
-        bs :: ByteString
-        bs = T.encodeUtf8 $ T.pack str
+utf8 = mkQuoter "utf8" (pure . T.encodeUtf8 . T.pack)
 
-        len :: Int
-        len = BS.length bs
+-- | Quasi-quoter to construct a /sized/ 'ByteString' literal. This quoter
+-- will parse Haskell character escapes, and interpret the quoted string as
+-- a sequence of bytes, rather than text.
+--
+-- @
+-- {-# LANGUAGE QuasiQuotes #-}
+--
+-- {- ... -}
+--
+-- let salt = [bytes|\xff\x00\xde|] :: 'SizedByteArray' 3 ByteString
+-- @
+--
+-- This uses Template Haskell to compute the length of the raw byte
+-- string you provide. Note that the string will be taken as is,
+-- with all whitespace preserved, for example @[bytes| x |]@ will have length 3.
+--
+-- The conversion rules are exactly the same as @OverloadedString@ rules for
+-- 'ByteString'. In particular, characters that are more than one byte wide
+-- will be silently truncated to one byte.
+bytes :: QuasiQuoter
+bytes = mkQuoter "bytes" (toBytes <=< parseEscapes)
+  where
+    toBytes = fmap pack . traverse charToByte
+    charToByte c =
+      let msg = "Character does not fit into a byte: '" <> [c] <> "' (" <> show c <> ")"
+      in maybe (fail msg) pure $ toIntegralSized (fromEnum c)
+    parseEscapes str = case reverse (readP_to_S (many lexChar) str) of
```

Comment:
Why is `reverse` needed here? I don’t think `ReadP` parsers can _actually_ return multiple results, and even if they could, why are we looking specifically at the last one?

If this `reverse` is really not important, then I suggest using `listToMaybe` to make the terrible `ReadS` interface a bit more readable.

Reply:
Ok, I see...

Reply:
The last one is the longest, yes, due to how `many` is implemented. `ReadP` parsers are a little weird in the regard they implement "completely fair disjunctions".

Since parsers are lazy, performance shouldn't be that bad here, AFAIG, it should still be linear in the length of the input.